src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs

```diff
@@ -462,30 +471,32 @@ public override int ReadBlock(Span<char> buffer)
         private void CompressBuffer(int n)
         {
             Debug.Assert(_byteLen >= n, "CompressBuffer was called with a number of bytes greater than the current buffer length.  Are two threads using this StreamReader at the same time?");
-            Buffer.BlockCopy(_byteBuffer, n, _byteBuffer, 0, _byteLen - n);
+            byte[] byteBuffer = _byteBuffer;
+            _ = byteBuffer.Length; // allow JIT to prove object is not null
+            new ReadOnlySpan<byte>(byteBuffer, n, _byteLen - n).CopyTo(byteBuffer);
```

Comment:
not related to the changes you have made: I can see that once the encoding is detected, this method is being called:

https://github.com/dotnet/runtime/blob/ff4b2451f6826aef886016fb9ab3ea21a5ada0b1/src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs#L491-L493

So for the default buffer size and Unicode:

https://github.com/dotnet/runtime/blob/ff4b2451f6826aef886016fb9ab3ea21a5ada0b1/src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs#L27

we copy 1022 bytes in place.

Why don't we just update `_byteLen` and `_bytePos`? I know it would require changing some other parts of the code that rely on the assumption that `_bytePos == 0` after the read:

https://github.com/dotnet/runtime/blob/ff4b2451f6826aef886016fb9ab3ea21a5ada0b1/src/libraries/System.Private.CoreLib/src/System/IO/StreamReader.cs#L654

I am asking this question because this PR improves the performance of `DetectEncoding` by moving rarely called code out of hot path and it seems like another perf improvement we could make here while we are at it.

Reply:
@stephentoub  just checking you saw this comment

Reply:
I did. It was preexisting this PR and my goal was just to land this PR. Anyone is welcome to follow up on the comment.